### PR TITLE
Migrate wasm.rs to test_zkasm.rs

### PR DESCRIPTION
**Pull request by akashin**
This PR moves code that generates ZK ASM from WASM module from `wasm.rs` to `test_wasm.rs` and runs it on a suite of testcases.

This will be a first step towards aligning our testing with the tooling that wasmtime uses. Ideally, we would convert this to tests similar to https://github.com/near/wasmtime/tree/main/cranelift/filetests/filetests/isa, but there are still a few non-trivial steps to be compatible with that setup (object generator implementation, text vs binary assembly), so starting with an intermediate solution.

### CODE REVIEW

File: cranelift/filetests/Cargo.toml [hunk 1]
```diff
@@ -10,7 +10,7 @@ publish = false
 edition.workspace = true
 
 [dependencies]
-cranelift-codegen = { workspace = true, features = ["disas"] }
+cranelift-codegen = { workspace = true, features = ["disas", "zkasm"] }
```
Comment (akashin) [inline cranelift/filetests/Cargo.toml:13]: @nagisa , this is the part about an additional feature I was asking about. I'm still trying to figure out why other targets don't need to enable their features explicitly...

Comment (nagisa) [reply cranelift/filetests/Cargo.toml:13]: They have got https://github.com/near/wasmtime/blob/a5e876765a938571cc1f097a78bb4ac5c5d84685/cranelift/Cargo.toml#L57 for example, and from there on I imagine it is again some sort of feature unification going on. FWIW I've just been running my tests with `cargo test --all-features` :shrug:

File: cranelift/filetests/src/test_zkasm.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    use cranelift_codegen::entity::EntityRef;
+    use cranelift_codegen::ir::function::FunctionParameters;
+    use cranelift_codegen::ir::ExternalName;
+    use cranelift_codegen::isa::zkasm;
+    use cranelift_codegen::{settings, FinalizedMachReloc, FinalizedRelocTarget};
+    use cranelift_wasm::{translate_module, DummyEnvironment};
+
+    fn generate_preamble(start_func_index: usize) -> Vec<String> {
+        let mut program: Vec<String> = Vec::new();
```
Comment (nagisa) [inline cranelift/filetests/src/test_zkasm.rs:13]:
```suggestion
        // TODO: preamble should be generated by a linker and/or clift itself.
        let mut program: Vec<String> = Vec::new();
```

File: cranelift/filetests/src/test_zkasm.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        program.push("start:".to_string());
+        program.push("  zkPC + 2 => RR".to_string());
+        program.push(format!("  :JMP(function_{})", start_func_index));
+        program.push("  :JMP(finalizeExecution)".to_string());
+        program
+    }
+
+    fn generate_postamble() -> Vec<String> {
+        let mut program: Vec<String> = Vec::new();
+        program.push("finalizeExecution:".to_string());
```
Comment (nagisa) [inline cranelift/filetests/src/test_zkasm.rs:23]: Out of curiosity, what is the point of this? From the presentation we've seen last week, IIRC Polygon folks use this to reset the registers back to their initial state or some such, but we seem to never do this really.

File: cranelift/filetests/src/test_zkasm.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        program.push("  ${beforeLast()}  :JMPN(finalizeExecution):".to_string());
+        program.push("                   :JMP(start)".to_string());
+        program
+    }
+
+    fn fix_relocs(
```
Comment (nagisa) [inline cranelift/filetests/src/test_zkasm.rs:29]: Could you also slap a `FIXME` on this? This test runner doesn't seem like a great place for relocation fixup and other such work. I am also suspicious of our need to have relocations at all!

File: cranelift/filetests/src/test_zkasm.rs [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            } else {
+                b"  UNKNOWN".to_vec()
+            };
+            delta += code.len() as i32;
+
+            code_buffer.splice(start..pos, code);
+        }
+    }
+
+    fn optimize_labels(code: &[&str], func_index: usize) -> Vec<String> {
```
Comment (nagisa) [inline cranelift/filetests/src/test_zkasm.rs:63]: Similarly we should probably look into producing the right code from the first principles. It is alright to have this here for the time being, but a `FIXME` would help to make sure this does not get lost.

File: cranelift/filetests/Cargo.toml (hunk 1 above)
Comment (akashin) [reply cranelift/filetests/Cargo.toml:13]: This fails with
```
error[E0432]: unresolved import `cranelift_codegen::isa::zkasm`
 --> cranelift/filetests/src/test_zkasm.rs:8:9
  |
8 |     use cranelift_codegen::isa::zkasm;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `zkasm` in `isa`
```

I think I'll leave this as is for now, but will add a comment to revise this part.

File: cranelift/filetests/src/test_zkasm.rs (hunk 2 above)
Comment (akashin) [reply cranelift/filetests/src/test_zkasm.rs:13]: I agree, I'm yet to find the exact place where this should be done.
There are https://docs.rs/cranelift-module/latest/cranelift_module/ and https://docs.rs/cranelift-object/latest/cranelift_object/ that seem to do something similar, but don't seem to be easily usable.

File: cranelift/filetests/src/test_zkasm.rs (hunk 4 above)
Comment (akashin) [reply cranelift/filetests/src/test_zkasm.rs:29]: I agree, added a TODO.

File: cranelift/filetests/src/test_zkasm.rs (hunk 5 above)
Comment (akashin) [reply cranelift/filetests/src/test_zkasm.rs:63]: I agree, label optimization should happen in `MachBuffer`, but we don't leverage it yet. Added a TODO

**Comment by akashin** [comment]
@nagisa  I've addressed comments and fixed tests, PTAL

**Review by nagisa — APPROVED** [review]

**Pull request merged**